docs/proposals/ssh_adapter.md

```diff
@@ -0,0 +1,194 @@
+# SSH protocol proposal
+
+We'd like to implement a protocol for Git LFS that uses SSH protocol
+exclusively, avoiding the need to use HTTPS altogether.  This will make
+deployment and use easier in a variety of situations where access to certain
+ports is limited.
+
+## What not to do
+
+There are several possible approaches that could be adopted.  SSH provides a
+native capability for the SFTP protocol, which can be used to transfer files.
+However, in order to implement this on the server side, each access (upload or
+download) must have an access control check instead of one at the beginning of
+the operation.  This would be inefficient in some server-side implementations,
+and nearly impossible to implement securely for implementations that use the
+system OpenSSH for implementation.
+
+## A more usable approach
+
+Git already has some places we can look for inspiration.  Its SSH protocol is
+based on the Git native protocol, which is based on the pkt-line scheme.
+Recently, Git has learned about protocol version 2, which provides better
+support for expressing and negotiating capabilities.
+
+Ideally, we would allow multiple operations to occur on a single connection for
+efficiency's sake, especially on high-latency connections, where the cost of SSH
+connection setup may be high due to multiple round trips.  In addition, a
+protocol which maps well onto HTTP may be beneficial for those server-side
+implementations which would like to proxy connections to an HTTP-based backend.
+
+This design assumes a reference to Git's pkt-line and protocol v2 documentation.
+pkt-line headers for this document may contain values up to 65519 decimal.
+
+To initiate a connection, Git LFS should run the folllowing command:
+
+    $ ssh [{user}@]{server} git-lfs-transfer {path} {operation}
+
+If authentication fails, or some other connection error occurs, errors will be
+read fronm standard error and displayed to the user.  The operation may be
+`upload`, `download`, or `meta` (for items which are not transfers).
```

Comment:
Can you explain what `meta` is?

Reply:
Yeah, I had intended to use it for locking and other cases where we're performing operations on the repo that are not strictly related to uploads or downloads, but that's probably not a good name.  If anyone has a better idea, I'm all ears.

Reply:
Hm, perhaps we could get rid of `meta` entirely, and specify `locking` instead, allowing any further extensions to be negotiated during the initial connection?

Reply:
I agree we should have an operation for locking instead.  It may be useful to have a way to discover capabilities without having to commit to an operation, but I don't feel strongly about it either way.